### PR TITLE
Refactor Tako

### LINKED ISSUE #3: Setup __health (and __gtg?)

### LINKED ISSUE #4: Set up Jest to use junit reporter in CI

### LINKED ISSUE #6: Set up app to send code coverage reports to Coveralls

### LINKED ISSUE #10: Create API endpoint to list repositories Tako is installed on
- [x] `GET` endpoint for retrieving list of repository names
- [x] `DELETE` endpoint for purging cached list of repository names
- [x] Convert existing code to TypeScript
- [x] Error handling
- [x] Unit tests

### LINKED ISSUE #21: Improve use of Prettier
- [ ] Decide how and when we want to apply Prettier (see [the docs](https://prettier.io/docs/en/))
- [ ] Fix `.editorconfig` for `*.md` files as it doesn't work with `prettier`

### PR DESCRIPTION
**Pull request by sjparkinson**
Best to not review this one side-by-side. Have a browse of the branch.

I've reworked most of @simonplend's logic, and have simplified the store of repositories down to a `Map`, for now.

~I'm not happy with having `app.repositoryStore` as a thing, but it does actually work.~

I did try adding to `context` using something like...

```js
const repositoryStore = new Map();

app.on('*', context => { context.repositoryStore = repositoryStore; });
```

... but that doesn't actually work, the modified context isn't available in other handlers for some reason.  Ideas welcome 💡.

There is an initial trio of tests, but I still need to write tests for...

- [x] `initialise.js`
- [x] `routes.js`

---

The idea is **not** to refresh on requests to the API, but to keep the list of repositories up-to-date by adding and removing them as we get the relevant events from GitHub. Building an initial list of repositories by calling `octokit.apps.getInstallationRepositories` when we start the application.

---

Resolves #3.
Resolves #4.
Resolves #10.
Resolves #6.

Kinda closes #21.

### CODE REVIEW

File: src/routes.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	// Get an express router from the Probot application instance
+	const router = app.route('/tako');
+
+	// Responses could frequently change, so send a sensible cache-control header.
+	router.use(noCache);
+
+	log.debug(`registered the noCache middleware`);
+
+	router.get('/repositories', (req, res) => {
+		res.send({
```
Comment (sjparkinson) [inline src/routes.js]: - [ ] This needs to read from `app.repositoryStore`.

File: src/initialise.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	const octokit = await app.auth().catch((err) => {
+		throw new Error('Failed to authenticate as a GitHub App', err);
+	});
+
+	logger.debug(`Authenticated as a GitHub App`);
+
+	/**
+	 * Get every installation of the GitHub App. There should only be one.
+	 *
+	 * Tako should be a private (also know as internal) GitHub App, installed on the same account it is administered by.
```
[the remark this answers is not in the document]
Comment (sjparkinson) [reply src/initialise.js]: Looking at https://developer.github.com/v3/apps/#get-the-authenticated-github-app, looks like this is possible with https://octokit.github.io/rest.js/#api-Apps-get.

```js
(await octokit.apps.get({})).owner.login
```

[the remark this answers is not in the document]
Comment (ker-an) [reply src/initialise.js]: I think I did it? Could you take a look at it? Any feedback is so welcome! I tested it out by creating K's Development Tako and installing it on a personal Node app.

File: src/initialise.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	assert(
+		Array.isArray(installations.data) && installations.data.length === 1,
+		'Tako should be an internal GitHub App, to configure this see https://developer.github.com/apps/managing-github-apps/making-a-github-app-public-or-private/#private-installation-flow.'
+	);
+
+	const installationAccount = installations.data[0].account.login;
+
+	const administratorAccount = (await octokit.apps.get({})).data.owner.login;
+
+	assert(installationAccount === administratorAccount, 'Tako should only be installed on accounts administered by financial-times-sandbox');
```
Comment (ker-an) [inline src/initialise.js]: @sjparkinson

[the remark this answers is not in the document]
Comment (ker-an) [reply src/initialise.js]: Done!

File: src/initialise.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	 *
+	 * This uses a private method on app, as we don't have an authenticated event context here.
+	 *
+	 * @see https://probot.github.io/api/latest/classes/application.html#auth
+	 * @see https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#authenticating-as-a-github-app
+	 *
+	 * @type {import('probot').GitHubAPI}
+	 */
+	const octokit = await app.auth().catch((err) => {
+		throw new Error('Failed to authenticate as a GitHub App', err);
```
[the remark this answers is not in the document]
Comment (sjparkinson) [reply src/initialise.js]: Added a custom error class that allows wrapping the underlying error.

**Comment by sjparkinson** [comment]
Stuck on https://github.com/probot/probot/issues/760, could sort out a workaround in `initalise.js` for now.

**Comment by sjparkinson** [comment]
@simonplend @quarterto this is now in a good place for a review.

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,26 +1,39 @@
...
       - checkout
 
-      # Download and cache dependencies
       - restore_cache:
           keys:
-          - v1-dependencies-{{ checksum "package.json" }}
-          # fallback to using the latest cache if no exact match is found
-          - v1-dependencies-
+            - npm-{{ arch }}-{{ .Branch }}-{{ checksum "package-lock.json" }}
+            - npm-{{ arch }}-{{ .Branch }}-{{ checksum "package-lock.json" }}
```
Comment (simonplend) [inline .circleci/config.yml]: Is `Branch` necessary? Seems like it'll just limit cache reuse.

File: .circleci/config.yml [hunk 6]
```diff
@@ -1,26 +1,39 @@
...
           paths:
-            - node_modules
-          key: v1-dependencies-{{ checksum "package.json" }}
+            - node_modules/
 
       - run: npm test
+
+      # Tried using npx, but there are weird issues when uploading.
+      - run:
+          name: codecov
```
Comment (simonplend) [inline .circleci/config.yml]: Why Codecov over Coveralls? Do we have subscriptions with both?

File: .circleci/config.yml [hunk 7]
```diff
@@ -1,26 +1,39 @@
...
+          command: |
+            npm install --save-dev codecov
```
Comment (simonplend) [inline .circleci/config.yml]: Why isn't this is a dev dependency in `package.json`?

File: .gitignore [hunk 8]
```diff
@@ -1,7 +1,82 @@
...
+.nuxt
+
+# vuepress build output
+.vuepress/dist
+
+# Serverless directories
+.serverless
+
+# FuseBox cache
+.fusebox/
```
Comment (simonplend) [inline .gitignore]: Nice! Where did this come from?

File: package.json [hunk 9]
```diff
@@ -28,28 +21,46 @@
...
     ]
   },
   "husky": {
     "hooks": {
       "pre-commit": "npm run lint"
     }
+  },
+  "prettier": {
+    "singleQuote": true,
+    "arrowParens": "always"
```
[the remark this answers is not in the document]
Comment (simonplend) [reply package.json]: These settings are to match what we do in Next (via the `.eslintrc.js` that `n-gage` puts in place), but I agree that may well not be what we want. I think "defeats the point" is overstating it a little, but we should certainly discuss this as part of https://github.com/Financial-Times/next/issues/233.

File: README.md [hunk 10]
```diff
@@ -13,27 +13,28 @@ A GitHub App that helps maintain a large number of repositories in a GitHub orga
...
-        "topics": [ "next-serverless" ]
+      "name": "next-fizz-buzz"
     }
-]
+  ]
+}
 ```
 
-`GET /tako/repositories?topic=next-app`
+`GET /tako/repositories/topic/serverless`
```
Comment (simonplend) [inline README.md]: Any reason for this API _not_ to be RESTful?

File: README.md [hunk 11]
```diff
@@ -61,10 +62,6 @@ npm run dev
 
 5. Probot will then automatically create you a `.env` file in your local working directory (you can add `LOG_LEVEL=trace` for more verbose logging)
 
-6. From your browser, copy the installiation ID from the URL (e.g. at `https://github.com/settings/installations/123456` the ID is `123456`)
-
-7. Add `TAKO_INSTALLATION_ID` to your `.env` file, setting it's value to the ID you've just copied
-
```
Comment (simonplend) [inline README.md]: :tada:

File: app.yml [hunk 12]
```diff
@@ -67,70 +67,48 @@ default_permissions:
...
-# The homepage of your GitHub App.
-# url: https://example.com/
-
-# A description of the GitHub App.
-# description: A description of my awesome app
-
 # Set to true when your GitHub App is available to the public or false when it is only accessible to the owner of the app.
-# Default: true
-# public: false
+public: false
```
Comment (simonplend) [inline app.yml]: Ahh, good that we can configure this in the manifest.

File: .circleci/config.yml (hunk 6 above)
Comment (sjparkinson) [reply .circleci/config.yml]: Simply because the GitHub App for Codecov was already installed.

File: .gitignore (hunk 8 above)
Comment (sjparkinson) [reply .gitignore]: My go to for `.gitignore`, https://github.com/github/gitignore/blob/master/Node.gitignore.

File: README.md (hunk 10 above)
Comment (sjparkinson) [reply README.md]: I went with what express supports.

Unfortunately I can't define a param as a separate route, so there'd be one route with two bits of logic.

Wasn't so sure about this one, and happy to change it, but wasn't sure how big the logic for filtering was going to be, so initially wanted to work on the two routes separately.

File: README.md (hunk 11 above)
Comment (sjparkinson) [reply README.md]: After this is merged, I'd like to spend an hour or so on documentation.

File: app.yml (hunk 12 above)
Comment (sjparkinson) [reply app.yml]: Yep!

Also turns out this is just a 1:1 mapping of the form inputs on https://github.com/settings/apps/new.

Still need to check that the following actually means an "internal" app like described elsewhere.

image

File: .circleci/config.yml (hunk 5 above)
Comment (sjparkinson) [reply .circleci/config.yml]: No, man is CircleCI complex.

File: package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (simonplend) [reply package.json]: Have discussed in person: we're going to go config-less with Prettier. Let's see how we go :man_dancing:

File: .circleci/config.yml (hunk 6 above)
Comment (simonplend) [reply .circleci/config.yml]: Ok, looks nicer than Coveralls anyway :grin:

File: .gitignore (hunk 8 above)
Comment (simonplend) [reply .gitignore]: Neat. We should consider this as part of creating a new app.

/me makes mental note for the future

File: app.yml (hunk 12 above)
Comment (simonplend) [reply app.yml]: Yay!

File: src/repositories.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Usage as below.
+ *
+ * Required properties are id (number), name (string), and topics (an array).
```
Comment (simonplend) [inline src/repositories.js]: Outdated comment: states that `topics` is required.

**Comment by simonplend** [comment]
What about `global.repositoryStore` instead? I'm not sure if we can trust `app`.

**Pull request merged**